### PR TITLE
Enhancements and Bug Fixes in WebSocket Consumers

### PR DESCRIPTION
**Pull request by amhajja**
## Key Changes:

- **New `RecordingConsumer` Class**: 
  - Implemented a new WebSocket consumer for handling recording functionalities.
  - Handles connection, disconnection, and message reception for recording data.
  - Includes functionality to accept or reject connections based on user authentication via token.
  - Sends transcript completion events to all clients in the group.

- **User Authentication via Token**: 
  - Introduced `get_user_from_token` function to fetch user info based on the provided token.
  - If no token is provided, the user is treated as an `AnonymousUser`.

- **Improved Message Handling**: 
  - Enhanced error handling for received messages to manage cases like missing fields and unknown message types.
  - Logging added for better traceability of received messages and errors.

- **Refactored Existing Consumers**: 
  - Removed unnecessary line breaks and simplified JSON dumps across existing consumers for better readability.
  - Consolidated error handling in existing functions to ensure consistency.

- **Environment-Specific Settings**: 
  - Modified settings to toggle debugging and channel layer configurations based on the environment (`development` vs `production`).
  - When in development mode, switched to `InMemoryChannelLayer` for easier testing.
  - Reverted to the original Redis channel layer for production to maintain performance and scalability.

- **Updated API Routing**: 
  - Added routing for the new `RecordingConsumer` in `api/routing.py` to ensure WebSocket connections can be established for recordings.

### CODE REVIEW

File: api/consumers.py [hunk 1]
```diff
@@ -30,9 +34,7 @@ async def connect(self):
         uc = await self.fetch_user_count()
 
         await self.send(
-            text_data=json.dumps(
-                {"type": "settings", "settings": settings, "user_count": uc}
-            )
+            text_data=json.dumps({"type": "settings", "settings": settings, "user_count": uc})
```
Comment (amhajja) [inline api/consumers.py:37]: Black fix

File: api/consumers.py [hunk 2]
```diff
@@ -109,9 +111,7 @@ async def delete_student(self, username):
     @database_sync_to_async
     def delete_student_from_db(self, username):
         session = QuizSession.objects.get(code=self.code)
-        student = QuizSessionStudent.objects.get(
-            quiz_session=session, username=username
-        )
+        student = QuizSessionStudent.objects.get(quiz_session=session, username=username)
```
Comment (amhajja) [inline api/consumers.py:114]: Black fix

File: api/consumers.py [hunk 3]
```diff
@@ -127,9 +127,7 @@ async def send_current_question(self):
         question_data = await self.fetch_current_question()
         if question_data:
             await self.send(
-                text_data=json.dumps(
-                    {"type": "current_question", "question": question_data}
-                )
+                text_data=json.dumps({"type": "current_question", "question": question_data})
```
Comment (amhajja) [inline api/consumers.py:130]: Black fix

File: api/consumers.py [hunk 4]
```diff
@@ -159,9 +157,7 @@ async def send_next_question(self):
         question_data = await self.fetch_next_question()
         if question_data:
             await self.send(
-                text_data=json.dumps(
-                    {"type": "next_question", "question": question_data}
-                )
+                text_data=json.dumps({"type": "next_question", "question": question_data})
```
Comment (amhajja) [inline api/consumers.py:160]: Black fix

File: api/consumers.py [hunk 5]
```diff
@@ -186,14 +182,10 @@ async def end_quiz(self):
             print("Failed to end the quiz; session not found.")
 
     async def start_quiz(self):
-        await self.channel_layer.group_send(
-            f"quiz_session_{self.code}", {"type": "quiz_started"}
-        )
+        await self.channel_layer.group_send(f"quiz_session_{self.code}", {"type": "quiz_started"})
```
Comment (amhajja) [inline api/consumers.py:185]: Black fix

File: api/consumers.py [hunk 6]
```diff
@@ -186,14 +182,10 @@ async def end_quiz(self):
...
 
         await self.send(
-            text_data=json.dumps(
-                {"type": "quiz_started", "message": "Quiz has started!"}
-            )
+            text_data=json.dumps({"type": "quiz_started", "message": "Quiz has started!"})
```
Comment (amhajja) [inline api/consumers.py:188]: Black fix

File: api/consumers.py [hunk 7]
```diff
@@ -215,9 +207,7 @@ def fetch_quiz_session(self):
 
     async def user_response(self, event):
         await self.send(
-            text_data=json.dumps(
-                {"type": "user_response", "response": event["response"]}
-            )
+            text_data=json.dumps({"type": "user_response", "response": event["response"]})
```
Comment (amhajja) [inline api/consumers.py:210]: Black fix

File: api/consumers.py [hunk 8]
```diff
@@ -285,9 +275,7 @@ def create_student_session_entry(self, username, code):
         try:
             session = QuizSession.objects.get(code=code)
             # studentUser = Student.objects.get(user_id=user_id)
-            student = QuizSessionStudent.objects.create(
-                username=username, quiz_session=session
-            )
+            student = QuizSessionStudent.objects.create(username=username, quiz_session=session)
```
Comment (amhajja) [inline api/consumers.py:278]: Black fix

File: api/consumers.py [hunk 9]
```diff
@@ -344,9 +332,7 @@ async def check_and_grant_skip_power_up(self, student_id):
             if session_settings.get("skip_question_logic") == "streak":
                 if (
                     student.skip_count < session_settings.get("skip_count_per_student")
-                    and correct_responses
-                    % session_settings.get("skip_question_streak_count")
-                    == 0
+                    and correct_responses % session_settings.get("skip_question_streak_count") == 0
```
Comment (amhajja) [inline api/consumers.py:335]: Black fix

File: hice_backend/settings.py [hunk 10]
```diff
@@ -243,6 +253,4 @@
 
 AWS_LAMBDA_INVOKER_ACCESS_KEY_ID = os.getenv("AWS_LAMBDA_INVOKER_ACCESS_KEY_ID")
 AWS_LAMBDA_INVOKER_SECRET_ACCESS_KEY = os.getenv("AWS_LAMBDA_INVOKER_SECRET_ACCESS_KEY")
-AWS_LAMBDA_INVOKER_REGION_NAME = os.getenv(
-    "AWS_LAMBDA_INVOKER_REGION_NAME", default="us-east-1"
-)
+AWS_LAMBDA_INVOKER_REGION_NAME = os.getenv("AWS_LAMBDA_INVOKER_REGION_NAME", default="us-east-1")
```
Comment (amhajja) [inline hice_backend/settings.py:256]: Black fix

**Review by aprask — APPROVED** [review]

**Pull request merged**